content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,530 @@
+  step_1()
+  
+  step_2()
+  
+  return(TRUE)
+}
+```
+
+Profiling tools would help us see that `step_1()` takes one second, while `step_2()` takes two. In practice, this is usually much harder to intuit visually. To profile code with profvis, use the `profvis()` function:
+
+```{r prof-slow-fn}
+result <- profvis(slow_function())
+```
+
+Printing the `result`ing object out will visualize the time different calls within `slow_function()` took:
+
+```{r print-slow-fn, eval = FALSE}
+result
+```
+
+![A screenshot of profvis output. A stack of grey bars sit atop a timeline that ranges from zero to three seconds. The bottom rectangle of the stack is labeled "slow_function" and stretches across the whole timeline. Two rectangles labeled "step_1" and "step_2" lie on top of the bottom rectangle, where the first stretches one-third of the way across the timeline and the second covers the remaining two-thirds.](slow-function-profvis.png)
+
+This output shows that, inside of `slow_function()`, `step_1()` took about a third of the total time and `step_2()` took two-thirds. All of the time in both of those functions was due to calling `pause()`.
+
+Profiling should be your first line of defense against slow-running code. Often, profiling will surface slowdowns in unexpected places, and solutions to address those slowdowns may have little to do with usage of tidy tools. To learn more about profiling, the [Measuring performance](https://adv-r.hadley.nz/perf-measure.html) chapter in Hadley Wickham's book [Advanced R](https://adv-r.hadley.nz/index.html) is a great place to start.
+
+### bench
+
+profvis is a powerful tool to surface code slowdowns. Often, though, it may not be immediately clear how to _fix_ that slowdown. The bench package allows users to quickly test out how long different approaches to solving a problem take.
+
+```{r load-bench, message = FALSE, warning = FALSE}
+library(bench)
+```
+
+For example, say we want to take the sum of the numbers in a list, but we've identified via profiling that this operation is slowing our code down:
+
+```{r numbers}
+numbers <- as.list(1:5)
+
+numbers
+```
+
+One approach could be using the `Reduce()` function:
+
+```{r reduce}
+Reduce(sum, numbers)
+```
+
+Another could involve converting to a vector with `unlist()` and then using `sum()`:
+
+```{r sum-unlist}
+sum(unlist(numbers))
+```
+
+You may have some other ideas of how to solve this problem! How do we figure out which one is fastest, though? The `mark()` function from bench takes in different proposals to solve the same problem and returns a tibble with information about how long they took (among other things.)
+
+```{r bm-reduce}
+res <-
+  mark(
+    approach_1 = Reduce(sum, numbers),
+    approach_2 = sum(unlist(numbers))
+  )
+
+res %>% select(expression, median)
+```
+
+The other nice part about `mark()` is that it will check that each approach gives the same output, so that you don't mistakenly compare apples and oranges.
```

Comment:
```suggestion
The other nice part about `bench::mark()` is that it will check that each approach gives the same output, so that you don't mistakenly compare apples and oranges.
```